plugins/outputs/cloudwatchlogs/internal/pusher/target_test.go

```diff
@@ -70,42 +71,83 @@ func TestTargetManager(t *testing.T) {
 
 		mockService := new(mockLogsService)
 		mockService.On("CreateLogStream", mock.Anything).Return(&cloudwatchlogs.CreateLogStreamOutput{}, nil).Once()
+		mockService.On("DescribeLogGroups", mock.Anything).Return(&cloudwatchlogs.DescribeLogGroupsOutput{
+			LogGroups: []*cloudwatchlogs.LogGroup{
+				{
+					LogGroupName:    aws.String(target.Group),
+					RetentionInDays: aws.Int64(0),
+				},
+			},
+		}, nil).Once()
 		mockService.On("PutRetentionPolicy", mock.Anything).Return(&cloudwatchlogs.PutRetentionPolicyOutput{}, nil).Once()
 
 		manager := NewTargetManager(logger, mockService)
 		err := manager.InitTarget(target)
+		assert.NoError(t, err)
+		// Wait for async operations to complete
+		time.Sleep(100 * time.Millisecond)
+		mockService.AssertExpectations(t)
+	})
+
+	t.Run("SetRetentionPolicy/NoChange", func(t *testing.T) {
+		target := Target{Group: "G", Stream: "S", Retention: 7}
+
+		mockService := new(mockLogsService)
+		mockService.On("CreateLogStream", mock.Anything).Return(&cloudwatchlogs.CreateLogStreamOutput{}, nil).Once()
+		mockService.On("DescribeLogGroups", mock.Anything).Return(&cloudwatchlogs.DescribeLogGroupsOutput{
+			LogGroups: []*cloudwatchlogs.LogGroup{
+				{
+					LogGroupName:    aws.String(target.Group),
+					RetentionInDays: aws.Int64(7),
+				},
+			},
+		}, nil).Once()
 
+		manager := NewTargetManager(logger, mockService)
+		err := manager.InitTarget(target)
 		assert.NoError(t, err)
+		time.Sleep(100 * time.Millisecond)
 		mockService.AssertExpectations(t)
+		mockService.AssertNotCalled(t, "PutRetentionPolicy")
 	})
 
 	t.Run("SetRetentionPolicy/LogGroupNotFound", func(t *testing.T) {
 		target := Target{Group: "G", Stream: "S", Retention: 7}
 
 		mockService := new(mockLogsService)
 		mockService.On("CreateLogStream", mock.Anything).Return(&cloudwatchlogs.CreateLogStreamOutput{}, nil).Once()
-		mockService.On("PutRetentionPolicy", mock.Anything).
-			Return(&cloudwatchlogs.PutRetentionPolicyOutput{}, &cloudwatchlogs.ResourceNotFoundException{}).Once()
+		mockService.On("DescribeLogGroups", mock.Anything).
+			Return(&cloudwatchlogs.DescribeLogGroupsOutput{}, &cloudwatchlogs.ResourceNotFoundException{}).Times(numBackoffRetries)
 
 		manager := NewTargetManager(logger, mockService)
 		err := manager.InitTarget(target)
-
-		assert.NoError(t, err) // The overall operation should still succeed even if setting retention policy fails
+		assert.NoError(t, err)
+		time.Sleep(30 * time.Second)
```

Comment:
Can we change the number of retries during the testing to make these shorter? Our unit tests already take a long time and I'd rather we didn't increase them by 1m30s with all these failure case checks. Can set numRetries as a field on the target manager that defaults to 5 and then overwrite that in the test.
```go
manager.numRetries = 2
```

Reply:
running long tests in parallel now